syntax = "proto3";

package gateway;

import "glossary.proto";
import "graph.proto";

service GatewayService {
  rpc GetTerm(glossary.GetTermRequest) returns (TermDetails);
  rpc GetTermByName(glossary.GetTermByNameRequest) returns (TermDetails);
  rpc SearchTerms(glossary.SearchTermsRequest) returns (SearchTermsResponse);
  rpc GetMindMapForTerm(GetMindMapForTermRequest) returns (GetMindMapForTermResponse);
  rpc AddTerm(glossary.AddTermRequest) returns (glossary.Term);
  rpc GetAllTerms(glossary.GetAllTermsRequest) returns (glossary.GetAllTermsResponse);
  rpc UpdateTerm(glossary.UpdateTermRequest) returns (glossary.Term);
  rpc DeleteTerm(glossary.DeleteTermRequest) returns (glossary.DeleteTermResponse);
  rpc AddRelationship(AddRelationshipRequest) returns (AddRelationshipResponse);
  rpc GetRelationshipsForTerm(graph.GetRelationshipsForTermRequest) returns (graph.GetRelationshipsForTermResponse);
  rpc DeleteRelationship(graph.DeleteRelationshipRequest) returns (graph.DeleteRelationshipResponse);
}

message TermDetails {
  glossary.Term term = 1;
  repeated graph.Relationship relationships = 2;
}

message SearchTermsResponse {
  repeated TermDetails results = 1;
}

message Node {
  string id = 1;
  string name = 2;
  string definition = 3;
}

message Edge {
  string from_id = 1;
  string to_id = 2;
  string label = 3;
}

message GetMindMapForTermRequest {
  string term_id = 1;
}

message GetMindMapForTermResponse {
  repeated Node nodes = 1;
  repeated Edge edges = 2;
}

message AddRelationshipRequest {
  string from_term_id = 1;
  string to_term_id = 2;
  graph.RelationshipType type = 3;
}

message AddRelationshipResponse {
  bool success = 1;
}